File: C++/ps_baekjoon/16724/main.cpp
//baekjoon ps 16724

#include <iostream>
#include <vector>
#include <queue>

using namespace std;
using pos = pair<int, int>;

vector<vector<pos>> uf;

pos find(pos data)
{
    int row = data.first;
    int col = data.second;
    if (uf[row][col] == data)
        return data;
    pos nextpos = find(uf[row][col]);
    return uf[row][col] = uf[nextpos.first][nextpos.second];
}

bool uni(pos a, pos b)
{
    a = find(a);
    b = find(b);
    if (a != b)
    {
        uf[a.first][a.second] = b;
        return true;
    }
    return false;
}

int main()
{
    ios_base::sync_with_stdio(false);
    int n, m;
    cin >> n >> m;
    int cnt = n * m;
    vector<vector<char>> map(n + 2, vector<char>(m + 2, 0));
    uf = move(vector<vector<pos>>(n + 2, vector<pos>(m + 2)));
    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= m; j++)
        {
            cin >> map[i][j];
            uf[i][j] = {i, j};
        }
    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= m; j++)
        {
            int nowrow = i;
            int nowcol = j;
            int nextrow;
            int nextcol;
            char direct = map[nowrow][nowcol];
            switch (direct)
            {
            case 'U':
                nextrow = nowrow - 1;
                nextcol = nowcol;
                break;
            case 'D':
                nextrow = nowrow + 1;
                nextcol = nowcol;
                break;
            case 'L':
                nextrow = nowrow;
                nextcol = nowcol - 1;
                break;
            case 'R':
                nextrow = nowrow;
                nextcol = nowcol + 1;
                break;
            default:
                break;
            }
            if (uni({nowrow, nowcol}, {nextrow, nextcol}))
                cnt--;
        }
    cout << cnt;
    return 0;
}

File: Java/ps_baekjoon/16724/Main.java
//baekjoon ps 16724

import java.util.*;
import java.util.stream.*;
import java.io.*;

class Pos {
    public int row, col;

    Pos(int row, int col) {
        this.row = row;
        this.col = col;
    }

    boolean isEqualTo(Pos obj) {
        return (this.row == obj.row) && (this.col == obj.col);
    }
}

public class Main {
    static int n, m;
    static int cnt;
    static int[] input;
    static ArrayList<String[]> map = new ArrayList<String[]>();
    static Pos[][] uf;

    static Pos find(Pos data) {
        int row = data.row;
        int col = data.col;
        if (uf[row][col].isEqualTo(data))
            return data;
        Pos nextPos = find(uf[row][col]);
        return uf[row][col] = uf[nextPos.row][nextPos.col];
    }

    static boolean uni(Pos a, Pos b) {
        a = find(a);
        b = find(b);
        if (a.isEqualTo(b))
            return false;
        else {
            uf[a.row][a.col] = b;
            return true;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        input = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        n = input[0];
        m = input[1];
        cnt = n * m;
        for (int i = 0; i < n; i++)
            map.add(Stream.of(br.readLine().split("")).limit(m).toArray(String[]::new));
        uf = new Pos[n][m];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                uf[i][j] = new Pos(i, j);
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++) {
                int nowRow = i;
                int nowCol = j;
                int nextRow = 0;
                int nextCol = 0;
                char direct = map.get(nowRow)[nowCol].charAt(0);
                switch (direct) {
                    case 'U':
                        nextRow = nowRow - 1;
                        nextCol = nowCol;
                        break;
                    case 'D':
                        nextRow = nowRow + 1;
                        nextCol = nowCol;
                        break;
                    case 'L':
                        nextRow = nowRow;
                        nextCol = nowCol - 1;
                        break;
                    case 'R':
                        nextRow = nowRow;
                        nextCol = nowCol + 1;
                        break;
                    default:
                        break;
                }
                if (uni(new Pos(nowRow, nowCol), new Pos(nextRow, nextCol)))
                    cnt--;
            }
        System.out.print(cnt);
    }
}